File: src/main/kotlin/org/tera201/vcsmanager/scm/SCMRepository.java
package org.tera201.vcsmanager.scm;

import lombok.Getter;

/* TODO Naming is confusing. */
public class SCMRepository {

	@Getter
	private String repoName;
	@Getter
    private String path; /* Path in local FS. */
	@Getter
    private String headCommit; /* Most recent commit. */
	@Getter
    private String firstCommit; /* First commit. */
	@Getter
    private SCM scm;
	private String origin; /* e.g. GitHub URL */

	public SCMRepository(SCM scm, String origin, String repoName, String path, String headCommit, String firstCommit) {
		this.scm = scm;
		this.origin = origin;
		this.repoName = repoName;
		this.path = path;
		this.headCommit = headCommit;
		this.firstCommit = firstCommit;
	}

    public String getOrigin() {
		return origin == null ? path : origin;
	}

	public String getLastDir() {
		String[] dirs = path.replace("\\", "/").split("/");
		return dirs[dirs.length-1];
	}

	@Override
	public String toString() {
		return "SCMRepository [path=" + path + ", headCommit=" + headCommit + ", lastCommit=" + firstCommit + ", scm="
				+ scm + ", origin=" + origin + "]";
	}

}

File: src/main/kotlin/org/tera201/vcsmanager/scm/SCMRepository.kt
package org.tera201.vcsmanager.scm

data class SCMRepository(
    val scm: SCM, // e.g. GitHub URL
    val origin: String,
    val repoName: String,
    val path: String, // Local file system path
    val headCommit: String, // Most recent commit
    val firstCommit: String // First commit
) {

    /** Returns the origin URL or the local path if origin is not provided. */
    fun getOrigin(): String = origin ?: path

    /** Returns the last directory in the path. */
    val lastDir: String
        get() {
            val dirs = path.replace("\\", "/").split("/".toRegex()).filter { it.isNotEmpty() }
            return dirs.lastOrNull() ?: ""
        }

    /** Provides a string representation of the SCM repository. */
    override fun toString(): String {
        return "SCMRepository(path='$path', headCommit='$headCommit', firstCommit='$firstCommit', scm=$scm, origin=$origin)"
    }
}